cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java

```diff
@@ -0,0 +1,228 @@
+                                               long earliestPutTs, InternalScanner s,
+                                               CompactionRequest request) throws IOException {
+    LOG.info("preCompact, filter using MessageDataFilter");
+    Scan scan = new Scan();
+    scan.setFilter(new MessageDataFilter(c.getEnvironment(), System.currentTimeMillis()));
+    return new StoreScanner(store, store.getScanInfo(), scan, scanners, scanType, store.getSmallestReadPoint(),
+                            earliestPutTs);
+  }
+
+  private Supplier<TransactionStateCache> getTransactionStateCacheSupplier(String tablePrefix, Configuration conf) {
+    String sysConfigTablePrefix = new HTable98NameConverter().getSysConfigTablePrefix(tablePrefix);
+    return new DefaultTransactionStateCacheSupplier(sysConfigTablePrefix, conf);
+  }
+
+  private static final class MessageDataFilter extends FilterBase {
+    private final RegionCoprocessorEnvironment env;
+    private final long timestamp;
+    private final TransactionVisibilityState state;
+
+    private byte[] prevTopicIdBytes;
+    private Long currentTTL;
+    private Integer currentGen;
+
+    MessageDataFilter(RegionCoprocessorEnvironment env, long timestamp) {
+      this.env = env;
+      this.timestamp = timestamp;
+      this.state = txStateCache.getLatestState();
+    }
+
+    @Override
+    public ReturnCode filterKeyValue(Cell cell) throws IOException {
+      int rowKeyOffset = cell.getRowOffset() + prefixLength;
+      int sizeOfRowKey = cell.getRowLength() - prefixLength;
+      long publishTimestamp = MessagingUtils.getPublishTimestamp(cell.getRowArray(), rowKeyOffset, sizeOfRowKey);
+      int topicIdLength = MessagingUtils.getTopicLengthMessageEntry(sizeOfRowKey) - Bytes.SIZEOF_INT;
+      int generationId = Bytes.toInt(cell.getRowArray(), rowKeyOffset + topicIdLength);
+
+      try {
+        if (prevTopicIdBytes == null || currentTTL == null || currentGen == null ||
+          (!org.apache.hadoop.hbase.util.Bytes.equals(prevTopicIdBytes, 0, prevTopicIdBytes.length,
+                                                      cell.getRowArray(), rowKeyOffset, topicIdLength))) {
+          prevTopicIdBytes = Arrays.copyOfRange(cell.getRowArray(), rowKeyOffset, rowKeyOffset + topicIdLength);
+          Map<String, String> properties = topicCache.get(ByteBuffer.wrap(prevTopicIdBytes));
+          currentTTL = Long.parseLong(properties.get(MessagingUtils.Constants.TTL_KEY));
+          currentGen = Integer.parseInt(properties.get(MessagingUtils.Constants.GENERATION_KEY));
+        }
+
+        // Old Generation (or deleted current generation) cleanup
+        if (MessagingUtils.isOlderGeneration(generationId, currentGen)) {
+          return ReturnCode.SKIP;
+        }
+
+        // TTL expiration cleanup only if the generation of metadata and row key are the same
+        if ((generationId == Math.abs(currentGen)) && ((timestamp - publishTimestamp) > currentTTL)) {
```

Comment:
No need to have `Math.abs` here. This is because if `currentGen < 0` and `dataGen == Math.abs(currentGen)`, it is already skipped by the previous `isOlderGeneration` condition.